### ISSUE TITLE
AndroidEnableProfiledAot not making any difference on startup time

### ISSUE THREAD

**Issue by claudioredi** [xamarin/xamarin-android#3791]
I'm testing `AndroidEnableProfiledAot` on my Xamarin.Android project (it's not Xamarin.Forms).

Unfortunately I haven't seen a single milisecond of improvement on startup time :( . I do see that building process is doing something since the apk size jumps from 62.5 mb to 79.2 mb (about 27% of increase) and I also see on build log that AOT is being performed.

I was really hoping this to make a difference so I would like to check if there is anything for me to validate to be sure AOT is working as you expected.

I'm seeing a log output that maybe rings any bell for you

`[monodroid] Unknown Mono AOT mode: 1`

My environment information

```=== Visual Studio Community 2019 for Mac ===

Version 8.3.1 (build 18)
Installation UUID: 7d6cc4c0-07de-400c-84f3-ad95ebd65b83
	GTK+ 2.24.23 (Raleigh theme)
	Xamarin.Mac 5.16.1.24 (d16-3 / 08809f5b)

	Package version: 604000198

=== Mono Framework MDK ===

Runtime:
	Mono 6.4.0.198 (2019-06/fe64a4765e6) (64-bit)
	Package version: 604000198

=== NuGet ===

Version: 5.3.0.6192

=== .NET Core SDK ===

SDK: /usr/local/share/dotnet/sdk/3.0.100/Sdks
SDK Version: 3.0.100
MSBuild SDKs: /Library/Frameworks/Mono.framework/Versions/6.4.0/lib/mono/msbuild/Current/bin/Sdks

=== .NET Core Runtime ===

Runtime: /usr/local/share/dotnet/dotnet
Runtime Versions:
	3.0.0
	2.1.13

=== Xamarin.Profiler ===

Version: 1.6.12.26
Location: /Applications/Xamarin Profiler.app/Contents/MacOS/Xamarin Profiler

=== Updater ===

Version: 11

=== Apple Developer Tools ===

Xcode 11.0 (14936)
Build 11A420a

=== Xamarin.Android ===

Version: 10.0.0.43 (Visual Studio Community)
Commit: xamarin-android/d16-3/8af1ca8
Android SDK: /Users/claudioredi/Library/Developer/Xamarin/android-sdk-macosx
	Supported Android versions:
		None installed

SDK Tools Version: 26.1.1
SDK Platform Tools Version: 28.0.2
SDK Build Tools Version: 29.0.2

Build Information: 
Mono: mono/mono/2019-06@7af64d1ebe9
Java.Interop: xamarin/java.interop/d16-3@5836f58
LibZipSharp: grendello/LibZipSharp/d16-3@71f4a94
LibZip: nih-at/libzip/rel-1-5-1@b95cf3f
ProGuard: xamarin/proguard/master@905836d
SQLite: xamarin/sqlite/3.27.1@8212a2d
Xamarin.Android Tools: xamarin/xamarin-android-tools/d16-3@cb41333

=== Microsoft Mobile OpenJDK ===

Java SDK: /Users/claudioredi/Library/Developer/Xamarin/jdk/microsoft_dist_openjdk_1.8.0.25
1.8.0-25
Android Designer EPL code available here:
https://github.com/xamarin/AndroidDesigner.EPL

=== Android SDK Manager ===

Version: 1.4.0.65
Hash: c33b107
Branch: remotes/origin/d16-3
Build date: 2019-09-19 20:42:44 UTC

=== Android Device Manager ===

Version: 1.2.0.115
Hash: 724ea69
Branch: remotes/origin/d16-3
Build date: 2019-09-19 20:43:06 UTC

=== Xamarin Designer ===

Version: 16.3.0.244
Hash: e15b2ee97
Branch: remotes/origin/d16-3
Build date: 2019-09-25 00:03:16 UTC

=== Xamarin.Mac ===

Xamarin.Mac not installed. Can't find /Library/Frameworks/Xamarin.Mac.framework/Versions/Current/Version.

=== Xamarin.iOS ===

Version: 13.2.0.42 (Visual Studio Community)
Hash: 5e8a208b
Branch: d16-3
Build date: 2019-09-17 14:19:30-0400

=== Xamarin Inspector ===

Version: 1.4.3
Hash: db27525
Branch: 1.4-release
Build date: Mon, 09 Jul 2018 21:20:18 GMT
Client compatibility: 1

=== Build Information ===

Release ID: 803010018
Git revision: 272ab73f00ab2b13ec97dc8b31d02980cf66d47f
Build date: 2019-09-27 19:26:51+00
Build branch: release-8.3
Xamarin extensions: 451b00e65052c11c16b79e6d31a24c4cf0d1b794

=== Operating System ===

Mac OS X 10.14.5
Darwin 18.6.0 Darwin Kernel Version 18.6.0
    Thu Apr 25 23:16:27 PDT 2019
    root:xnu-4903.261.4~2/RELEASE_X86_64 x86_64

```
Please let me know any information I could provide for better troubleshooting this.

**Comment by radekdoulik** [comment]
Could you please attach the diagnostic build log?

How do you measure the startup time?

**Comment by claudioredi** [comment]
@radekdoulik : thanks for your quick response. Attached the diagnostic log.

[output.txt](https://github.com/xamarin/xamarin-android/files/3730438/output.txt)

For now my measure is done with a manual stopwatch execution. I measure how much time I see the splash screen. Since I expect a noticiable reduction of the startup time, I thought this could be enough for an initial test. It takes 5.5 seconds on the device I'm currently testing and it doesn't change with `AndroidEnableProfiledAot` true or false.

**Comment by radekdoulik** [comment]
Could you please try newer release? https://github.com/xamarin/xamarin-android/releases/tag/v10.0.3.0 There were some AOT performance related fixes and I am not sure, whether they were already included in your version.

You might try to check the `adb logcat` output for activity manager messages, like

    10-15 11:56:41.134  1318  1360 I ActivityTaskManager: Displayed com.microsoft.smarthotel/crc6408526390d3d53237.MainActivity: +1s936ms

which shows the time in ms. In the older versions it will contain `ActivityManager` instead of `ActivityTaskManager`

What device/abi do you use?

You might also try normal AOT (`AotAssemblies=true` without enabling the profiled AOT) to see whether AOT improves the speed on your device.

**Comment by claudioredi** [comment]
The device is a Pixel 3a (arm64)
Using logcat I see this with `AndroidEnableProfiledAot` enabled

`2019-10-15 22:26:14.964 1147-1201/? I/ActivityManager: Displayed com.myapp.myappdev/com.myapp.SplashScreen: +1s506ms`

and this when I disabled it

`2019-10-15 22:26:14.964 1147-1201/? I/ActivityManager: Displayed com.myapp.myappdev/com.myapp.SplashScreen: +1s477ms`

This time is not representing the visual startup time though (time for the splash screen to disappear). 

Tomorrow will test normal AOT.

**Comment by claudioredi** [comment]
Full AOT fails on runtime for me. Mvvmcross framework fails to resolve a class... I already tested AOT in the past and I got the same crash. I'm afraid for now I'm not going to be able to compare normal AOT versus profiled.

Let you know if I get that working so we can compare startup times of normal AOT vs profiled.

**Comment by radekdoulik** [comment]
If possible, you might try the newer version with profiled AOT. Hopefully it would improve the performance for you.

**Comment by Cybrosys** [comment]
Did a quick test with our slowest enterprise hardware where we have an app (Xamarin Android + Mvx) which takes 22 seconds from touching the icon to when the splash screen goes away. Changing the AndroidEnableProfiledAot value makes no difference. On a more modern hardware the app boots up in around 3 seconds regardless of the value as well.

I haven't reviewed what exactly is taking time, maybe it's the Mvx/MvvmCross IoC configuration or something else,.

**Comment by nll** [comment]
Recently I also try to enable the `AndroidEnableProfiledAot` on a Xamarin Android application and it did get any performance improvement and just an increase in the apk size (using Xamarin.Android 10.1.1.0).

Only when creating a custom profile following [this](https://github.com/xamarin/xamarin-android/blob/a01df1b4db775b940a5a40f44f16cd58fdb448be/Documentation/guides/profiling.md#profiling-the-aot-compiler) instructions and added something like this to the project:

```csproj
  ...
  <PropertyGroup>
    <AndroidEnableProfiledAot>True</AndroidEnableProfiledAot>
    <AndroidUseDefaultAotProfile>False</AndroidUseDefaultAotProfile>
  </PropertyGroup>
  <ItemGroup Condition="'$(AndroidUseDefaultAotProfile)' == 'False'">
    <AndroidAotProfile Include="$(MSBuildThisFileDirectory)custom.aprof" />
  </ItemGroup>
  ...
```
I got the performance boost on the app startup that I was looking for.

**Comment by bellissimo** [comment]
@nll That is remarkable, thanks for pointing this out. I just went through the process and startup time went from 2.1s with AndroidEnableProfiledAot turned on, to 1.4s with the custom profile generated. In fact it startup up slightly quicker than having AotAssemblies turned on, which is a bit weird. The resulting APK size with the custom profile is only slightly larger than with AndroidEnableProfiledAot, and quite a bit smaller than full AOT.

It begs the question, why is this not the default setup over AndroidEnableProfiledAot? Is there a drawback I am missing, or may it vary per app?

**Comment by radekdoulik** [comment]
The recording of own AOT profile is a feature we are still working on, to make it as simple as possible. Context: https://github.com/xamarin/xamarin-android/commit/853b367ca8a4bd4ecb2a9bbf4de2937397c9200d

The startup with own profile is indeed usually slightly faster than using just `AotAssemblies=true` as there's less I/O during startup and also the "hot code" is located together.

**Comment by justinwojo** [comment]
I've been attempting to get the above recommendations to work on a Mac without success. I have enabled the `AndroidEmbedProfilers` property, verified that it's actually working and running in the verbose build output, set the `debug.mono.profile` to `aot:port=9999` (validated it was set by calling getprop as well), started the app and waited 15 seconds or so, ran the `aprofutil -s -v -f -p 9999 -o myprofile.aprof` command, but always getting an error that there is no data:

```
Calling 'adb forward tcp:9999 tcp:9999'...
Reading from '127.0.0.1:9999'...
Read total 0 bytes...
Error: aotprofile-tool: Unable to read profile through local port: 9999.
System.IO.IOException: Input file is too small.
```

I've tried rebuilding, changing project settings like Shared Mono Runtime on and off, etc. Cannot seem to get it to record anything. Any suggestions would be great, or we'll patiently wait for the above commit to be available and see if it works then (even though it's all the same commands).
